pysingfel/gpu/diffraction.py

```diff
@@ -78,6 +93,18 @@ def calculate_diffraction_pattern_gpu(reciprocal_space, particle, return_type='i
         pattern_cos, pattern_sin, atom_type_num, cuda_split_index,
         pixel_number)
 
+    # Add the hydration layer
+    water_position = np.ascontiguousarray(particle.mesh[particle.solvent_mask,:])
+    water_num = np.sum(particle.solvent_mask)
+    water_prefactor = 0.334 * 10**30 * particle.mesh_voxel_size**3
+
+    cuda_water_position = cuda.to_device(water_position)
+
+    calculate_solvent_pattern_gpu_back_engine[(pixel_number + 511) // 512, 512](
```

Comment:
do we have to hard-code 512 pixels here?

Reply:
I don't know much about this...
Looking into it a little bit, it sounds like it could be benchmarked: see [this](https://docs.nvidia.com/cuda/cuda-c-programming-guide/#occupancy-calculator) and [this](https://docs.nvidia.com/cuda/cuda-occupancy-calculator/index.html) from NVIDIA, or section "Choosing a block size" in [this course](https://nyu-cds.github.io/python-numba/05-cuda/) from NYU.

@AntoineDujardin do you think we would benefit from setting this as a variable for benchmarking?

Reply:
That's not the number of pixels and that has nothing to do with the `(4, 512, 512)` shape of the detector. That's the number of threads and warps used by CUDA. We need the product of the two to be at least equal to the number of pixels, hence the `//512` (with `+511` for rounding up) and `512`.
We might be able to have better performances with another number than 512, but the optimal number might not be the same on all the machines and in all the `..._gpu_back_engine` functions. I'd keep 512 until we address performances specifically.

Reply:
Right, I was thinking of some machine-specific set of parameters like this one (might be the only one, dunno), but I agree this can wait

Reply:
ok, let's keep it like this for now.